### PR TITLE
Add package `Cartopy`

### PR DESCRIPTION
**Pull request by juntyr**
### Description

`Cartopy` is a commonly used library to draw maps with `matplotlib`. Since most of its dependencies are already in `pyodide`, except for the pure Python dependency `pyshp`, adding `cartopy` turned out to be quite simple.

`Cartopy` often downloads shape files the first time that the user requires them during plotting, which is only supported if the `pyodide-http` patches are applied. How should this be handled?

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests
- [ ] Add new / update outdated documentation

### CODE REVIEW

**Comment by juntyr** [comment]
Mmmm, I'm not sure if I understand why the tests fail - could someone perhaps give me a pointer in the right direction? Thanks :)

File: packages/pyshp/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package:
+  name: pyshp
+  version: 2.3.1
+  top-level:
+    - shapefile
+source:
+  url: https://files.pythonhosted.org/packages/98/2f/68116db5b36b895c0450e3072b8cb6c2fac0359279b182ea97014d3c8ac0/pyshp-2.3.1-py2.py3-none-any.whl
+  sha256: 67024c0ccdc352ba5db777c4e968483782dfa78f8e200672a90d2d30fd8b7b49
+test:
```
Comment (hoodmane) [inline packages/pyshp/meta.yaml:9]:
```suggestion
requirements:
  run:
    cartopy
test:
```

Comment (juntyr) [reply packages/pyshp/meta.yaml:9]: Thanks, that got me onto the right track

**Comment by juntyr** [comment]
@hoodmane Ok, I've made some progress and may need your help again with decyphering and fixing the remaining errors. Can `matplotlib` be used in a test or are there some restrictions I should be mindful of?

**Review by ryanking13 — COMMENTED** [review]
Thanks @juntyr!

> Cartopy often downloads shape files the first time that the user requires them during plotting, which is only supported if the pyodide-http patches are applied. How should this be handled?

I think using `pyodide-http` makes sense.

> Can matplotlib be used in a test or are there some restrictions I should be mindful of?

You can see [matplotlib tests](https://github.com/pyodide/pyodide/blob/main/packages/matplotlib/test_matplotlib.py) for a reference when writing tests that use matplotlib. 
You'll need to add some tests that check if the images (plots) are correctly created in Pyodide. 

For example, you can run the code that generates an image (plot) locally, save the image into a file, then compare it with the same image created from Pyodide. You can copy the helper functions from matplotlib tests.


```
conftest.py:202: in pytest_runtest_call
    yield from extra_checks_test_wrapper(browser, trace_hiwire_refs, trace_pyproxies)
conftest.py:233: in extra_checks_test_wrapper
    assert (delta_proxies, delta_keys) == (0, 0) or delta_keys < 0
E   assert ((16, 34) == (0, 0)
```

You can skip this pyproxy check by adding the following decorators (again, check matplotlib tests to see how these decorators are applied). The error means there are some leaked Python objects after the test and it is quite normal when testing a package that interacts with JS.

```
pytest.mark.skip_refcount_check
pytest.mark.skip_pyproxy_check
```

File: packages/libproj/patches/0001-stod-empty-zero.patch [hunk 2]
```diff
@@ -0,0 +1,16 @@
+diff --git a/src/iso19111/internal.cpp b/src/iso19111/internal.cpp
```
[the remark this answers is not in the document]
Comment (juntyr) [reply packages/libproj/patches/0001-stod-empty-zero.patch:1]: Before https://github.com/pyodide/pyodide/pull/3909/commits/9a378a8098a63e98e2225a5733510db0d5d2f28b, the tests fail with a C++ invalid argument exception in this string to double parsing function, which is not triggered in a non-WASM build. I have no idea if the implementation of double-from-string parsing for an empty string is different in emscripten than elsewhere or if the empty string case is only triggered in WASM for some reason. However, explicitly handling `""` -> `0.0`, which is what C's `strtod` does, fixes the issue. I found this fix a few weeks back after about a day of randomly trying some ideas (and just forgot to include it in the initial PR commit). Perhaps we can find a better fix? For now this seems to work and allow `cartopy` to load.

[the remark this answers is not in the document]
Comment (juntyr) [reply packages/libproj/patches/0001-stod-empty-zero.patch:1]: Should I add that comment into the patch source code or the `meta.yaml` file?

[the remark this answers is not in the document]
Comment (ryanking13) [reply packages/libproj/patches/0001-stod-empty-zero.patch:1]: Okay, thanks for the explanation! Maybe the problem is related to the locale but I am not sure. Anyway, your fix looks reasonable to me.

> Should I add a short explanation into the patch source code or the meta.yaml file?

You can add a comment directly to the patch file.

File: packages/pyshp/test_pyshp.py
Comment (ryanking13) [inline packages/pyshp/test_pyshp.py:1]: We run import tests for all packages by default so this is redundant. I think it is okay to remove it.

**Review by ryanking13 — APPROVED** [review]
Thanks. @juntyr! Please update the changelog.

File: packages/pyshp/test_pyshp.py
Comment (juntyr) [reply packages/pyshp/test_pyshp.py:1]: I've removed the test in [b8190e5](https://github.com/pyodide/pyodide/pull/3909/commits/b8190e58d307512d6f0f6c112f7da8695814e061).

File: packages/libproj/patches/0001-stod-empty-zero.patch (hunk 2 above)
[the remark this answers is not in the document]
Comment (juntyr) [reply packages/libproj/patches/0001-stod-empty-zero.patch:1]: I have done so in [b8190e5](https://github.com/pyodide/pyodide/pull/3909/commits/b8190e58d307512d6f0f6c112f7da8695814e061).

**Comment by ryanking13** [comment]
Thanks again @juntyr!

**Comment by juntyr** [comment]
Thanks for the help and reviews @ryanking13 and @hoodmane!

**Pull request merged**